### ISSUE TITLE
Problems with multiple encodes

### ISSUE THREAD

**Issue by wrender** [nebhead/PlexPostProc#15]
I'm having an issue with Plex and this script (the ffmpeg one). When I try and record 1 DVR channel this works.  When I try to simultaneously record 2 different DVR channels, the second recorded episode folder has no file.  Plex gives an error saying silly script deleted the file.

I can see that both files are being written to by Plex in the .grab folder, and then when the first file is converted and written it seems to kill off the second ffmpeg as the .grab file for the second recording just disappears.

**Comment by nebhead** [comment]
Ok, interesting.  I'm not sure I have the same issue here, but I rarely have two recordings going at the same exact time.  Are the recordings starting and ending at the same time?  Which branch of the script are you using, the FFMPEG or the Handbrake version?

**Comment by wrender** [comment]
I'm using the FFMPEG. But I believe I tested it with HandBrake-CLI and had the same issue. No. The recordings don't have to start or end at the same time. But they need to overlap. For example two FFMPEG processes running at the same time.

**Comment by wrender** [comment]
The easiest way I found to test was just go into the DVR Guide, and click to record two different shows, then click cancel on each of them. When you cancel them they execute the post process script.
Then just watch the destination folder you set with something like watch -n 1 ls -la /.grab

**Comment by nebhead** [comment]
Ok, let me give it a shot today after work and see if I can reproduce.  Also, what version of Plex are you currently on?  What OS?  

I'm personally using the Official Docker, under Ubuntu 16.04 Server Edition.

**Comment by wrender** [comment]
Ok. Awesome. I'm curious to see if you can re-produce the issue.
I'm running CentOS 8, Docker-ce, and the Official Docker. I just updated to 1.18.9.2578, today but haven't had a chance to test the script again.  I did just test the built in DVR transcoder in Plex and it works with two recordings at a time.  The main reason I want to use a script, is so that I can delay the transcoding. It is an older server and CPU can't keep up with 3 trancodes at a time.

**Comment by nebhead** [comment]
Hmm... I also updated today and I'm at Version 1.19.1.2589.  Regardless, I'll give it the old college try and see what happens.

**Comment by nebhead** [comment]
Yeah, tried removing the RM command just to be sure that wasn't causing any issues and replaced with a MV command.  It seems like for whatever reason, the temporary folder that is created in the .grab folder for one of the recordings gets clobbered.  I think Plex might be doing this, not expecting multiple scripts to be running simultaneously.  I think we are going to need to open a thread in the Plex forums.  :(  I'll get something going.

**Comment by nebhead** [comment]
Workaround might be to implement a lock file, like [this script,](https://thatvirtualboy.com/2017/11/28/plex-dvr-postprocessing-script.html) so that only one is processing at a time.  Seems like that might be what you want anyway, given your comments above.  If that would fix things for you, I'll implement that and update the repo.

**Comment by wrender** [comment]
Ok. Ya. I tried taking the components out of thatvirtualboy script that create a lock file, and a while loop to stagger the transcoding but the issue still happens. So I think you are right, must be something on the plex side of things....

**Comment by nebhead** [comment]
Yep - I posted to the Plex Forum and I hope we can get someone to take some interest there.  I suspect that the Plex server gets a little over zealous with garbage clean-up after the first script completes.

**Comment by nebhead** [comment]
FYI, here's the link to the Plex Forum issue/question that I raised: https://forums.plex.tv/t/simultaneous-post-processing-scripts-fail/565716 

The last person who raised this didn't get any response, so I'm not super confident.  Maybe if you jump on there and add to the discussion someone will take notice?  I'm just guessing their algorithms are might see more than one person commenting on a thread and assume it's potentially real.

**Comment by nebhead** [comment]
FYI:  I updated the FFMPEG branch with a lockfile implementation that I think fixes the issue.  I've tested it and it appears to work for me.  Basically what it does is create a lock file in the /tmp folder when it kicks of the transcode, then when finished, it deletes the lock file.  Then it checks to see if there are any other lock files in the /tmp directory before exiting the script.  Basically this is just holding all processes until they are all done.  This way, Plex doesn't get to trample on the .grab folders until all scripts are finished.  Feel free to give it a try.  I'm going to go ahead and close this thread, but do let me know if you see further issues.

**Comment by codeboyQ2n5ha45** [comment]
I tested it on my end. It worked initially, and then I stopped an started a few simultaneous recordings and it stopped working and only writes the second recording.  I think it needs work.  For now I might just turn of transcoding, and manually cron job convert them at night.

**Comment by wrender** [comment]
I tested it on my end. It worked initially, and then I stopped an started a few simultaneous recordings and it stopped working and only writes the second recording. I think it needs work. For now I might just turn of transcoding, and manually cron job convert them at night.

**Comment by codeboyQ2n5ha45** [comment]
Ok. I'm testing it again now too... I'm seeing what happens when I don't manually click "Cancel recording" and just let them go to the end of the recording.

**Comment by wrender** [comment]
Ok. I'm testing it again now too... I'm seeing what happens when I don't manually click "Cancel recording" and just let them go to the end of the recording.

**Comment by wrender** [comment]
I agree though, it would be great if the problem was fixed/clarified on the Plex side of things. Do you think there would be an easy way to recursively convert videos in a folder in the middle of the night? Like some sort of cron job?  That way I would just leave them as mpeg2 or whatever they come in from HDhomerun as, and then later convert them to store disk space.

**Comment by nebhead** [comment]
So, I'm probably more of a Python guy than a Bash guy, so I would probably run a CRON job with Python that scanned my media folders for any new files (i.e. files that were recorded that day), store those in a dictionary/list, and then one by one encode them with FFMpeg serially.  

If your video card is good enough, maybe you can try to use the hardware encoding capabilities of the card to offload the CPU?

**Comment by codeboyQ2n5ha45** [comment]
Interesting. I just have a cheap GeForce GT 710 in the server plex is on. So not sure if it is good enough for hardware encoding. I did get it working for hardware decoding...

**Comment by wrender** [comment]
Interesting. I just have a cheap GeForce GT 710 in the server plex is on. So not sure if it is good enough for hardware encoding. I did get it working for hardware decoding...

**Comment by nebhead** [comment]
Oh yeah, that's probably not going to work.  You need a pretty beefy nVidia card to get the hardware transcoding.  What CPU do you have?  Is it an Intel?  You might still be able to use Intel Quick Sync Video Hardware Transcoding.

**Comment by wrender** [comment]
Thanks. It's an old Intel(R) Xeon(R) CPU X3450 @ 2.67GHz, 8 cores. But those slow old cores fill up quick :)  And no Intel Quick Sync unfortunately. I'll have to upgrade soon...

**Comment by janlavoie** [comment]
TL;DR: Thanks for the script! The workaround using ppplock files didn't work for me on mac for whatever reason. I ended up working around the plex issue using hard links and an expectation that plex would delete the files/folder. Working well so far. Figured maybe you'd be interested.

Cheers!


@nebhead HI! Sorry to post on an old bug, but I wanted to say I ran into this plex issue as well (I filled https://forums.plex.tv/t/bug-pms-concurrent-dvr-postprocessing-fails-because-grab-uuid-folder-of-second-run-is-deleted/918187 with details, although it's probably the same issue as the one you filled in 2020).

I'm running an ancient version of your script to post-process (slightly modified). 

I tried upgrading to the latest version of your script, but the workaround that is implemented with the ppplock files wasn't really working for me (I suspect perhaps some platform differences with mktemp on macos, or some such, but it's unclear whether it's that or something else). I did try troubleshooting it and fixing it, but I wasn't really getting anywhere with the locking and delayed-exit-strategy...

After troubleshooting some more and thinking about it, I approached the problem differently : Instead of delaying the exit of the script if more than one instance is running, I just assumed plex would delete stuff, and used a strategy involving hard links and a work folder. This probably doesn't work for everyone, since it does require hard link support (I mean you could copy the files I guess instead, but that might be tedious/long... anywho, I digress)

Assuming the filesystem you're using supports hard links, you can hard link the source file somewhere else, generate the converted file outside of plex's .grab/uuid folder, and then when you're done, re-create the .grab/uuid folder and hard link the converted file there.

e.g (pseudo bash-ish, possibly full of mistakes (like missing quotes at the very least))

```

# Setup some variables 

$SOURCEFILE=$1
$SOURCEDIR=`dirname $SOURCEFILE`

$WORKDIR=$SOURCEDIR/../.work/$$ # or whatever. probably needs to be on the same filesystem as the .grab folder
$WORKSOURCEFILE=$WORKDIR/`basename $SOURCEFILE`
$WORKDESTFILE=${WORKSOURCEFILE%.ts}.mkv
$DESTFILE=$SOURCEDIR/`basename $WORKDESTFILE`

# Create work folder
mkdir -p $WORKDIR

# Part 1 - hard link ts file to work folder
ln $SOURCEFILE $WORKSOURCEFILE

# Part 2 - Convert hard linked input to output in work folder
convertWhatever -i $WORKSOURCEFILE -o $WORKDESTFILE


#Part 3 - Recreate plex's folder, just in case it deleted it because reasons
mkdir -p $SOURCEDIR

#Part 4 - Hard link converted shiny file into plex's folder
ln $WORKDESTFILE $DESTFILE

# Plex could still delete stuff, so this script needs to end as soon as possible
# You could retry part 3 and 4 here, but since we probably can't eliminate the race condition anyway, it's probably useless

# Part 5 - cleanup and GTFO as soon as possible - plex could delete stuff at anytime probably...
if [ -f  $DESTFILE ] ; then 
   #seems like it worked maybe... do Le Cleanup
   rm $WORKSOURCEFILE
   rm $WORKDESTFILE
   rm $SOURCEFILE
   # rmdir $WORKDIR # should be empty... probably safer? or rm -r #YOLO 
else
   #oops, leave stuff there or whatever, log, call ghostbusters. 
fi
```


Long story short, I found a different workaround that worked for me, so I figured I'd share it in case it's useful to you or anyone at some point. It's still probably racy, but I've had good success with it so far. 

In any case, thanks for 1) reading this far, 2) sharing your script. Much appreciated!

Cheers,
